### PR TITLE
kubeadm: simplify discover/token and add detailed unit tests

### LINKED ISSUE #1917: refactor discovery/token/token.go
this PR fixed an issue where token based join can block indefinitely if the server is not reachable:
https://github.com/kubernetes/kubernetes/pull/85156

the logic in the file can be simplified a lot and the wait group is not needed.
WIP: https://github.com/neolit123/kubernetes/commit/bbbc467e7897c20bed3a3053790d1f87fa891ba9

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:
- Add retrieveValidatedConfigInfo to be able to better unit
test the logic in token.go
- Break some of the logic in RetrieveValidatedConfigInfo into
helper functions.
- Pass JoinConfiguration.Discovery to RetrieveValidatedConfigInfo
instead of JoinConfiguration.
- Use the discovery timeout per API call to fetch cluster-info
(optionally the user value can be slit in 2). This removes
the need for a wait group.
- Add detailed unit tests for retrieveValidatedConfigInfo.

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#1917

**Special notes for your reviewer**:

details about the refactor:
https://docs.google.com/document/d/1lf-DsWoJiZAXm9iVeO4MoP0ku0z8S30Izm9uLYigaUw/edit#heading=h.ee22he69cqq

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/priority important-longterm
/kind cleanup
@kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

**Review by rosti — COMMENTED** [review]
Thanks @neolit123 !
A much needed change. However, I think that we can simplify it even further by moving the timeout management away from the actual API calls and into a central call (in `fetchKubeConfigWithTimeout`). We can increase the retry interval size and add some jitter in the retries so that we won't nuke the API endpoint if multiple nodes are joining at the same time.

**Comment by neolit123** [comment]
@rosti
> A much needed change. However, I think that we can simplify it even further by moving the timeout management away from the actual API calls and into a central call (in fetchKubeConfigWithTimeout).

could you please clarify with examples what do you mean by moving the timeouts away from the API calls and "central call"?

after the change, the API calls are moved into a function that performs a poll.
fetchKubeConfigWithTimeout was wrapping the whole API object fetch and verification logic in a poll, which is far from ideal.

> We can increase the retry interval size and add some jitter in the retries so that we won't nuke the API endpoint if multiple nodes are joining at the same time.

i'm ok with changing the PollImmediate to a exp. backoff with jitter.

**Comment by rosti** [comment]
> could you please clarify with examples what do you mean by moving the timeouts away from the API calls and "central call"?
 Keep `fetchKubeConfigWithTimeout` and just change the way it retries + no retries in the actual API calls.

> i'm ok with changing the PollImmediate to an exp. backoff with jitter.
> but given the timeout is a user value, i guess we should manually apply the jitter with time.Sleep()

I think, that there are some jitter funcs without the exponential backoff (see `wait.JitterUntil` and the `wait.Jitter*` funcs).

**Comment by neolit123** [comment]
> Keep fetchKubeConfigWithTimeout and just change the way it retries + no retries in the actual API calls.

i think the retries should be only on the API calls. the validation does not need retries - this is a problem in fetchKubeConfigWithTimeout.

> I think, that there are some jitter funcs without the exponential backoff (see wait.JitterUntil and the wait.Jitter* funcs).

i will check these utils and update.

**Comment by neolit123** [comment]
>> I think, that there are some jitter funcs without the exponential backoff (see wait.JitterUntil and the wait.Jitter* funcs).

> i will check these utils and update.

@rosti i've added usage of `wait.JitterUntil`

File: cmd/kubeadm/app/discovery/token/token.go [hunk 1]
```diff
@@ -174,42 +150,85 @@ func buildSecureBootstrapKubeConfig(endpoint string, caCert []byte, clustername
...
+
+	// Create client from kubeconfig
+	if client == nil {
+		client, err = kubeconfigutil.ToClientSet(kubeconfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), duration)
```
Comment (rosti) [inline cmd/kubeadm/app/discovery/token/token.go]: The context here gives me an idea. What about creating a root context (at the beginning of `retrieveValidatedConfigInfo`), that contains all the timeout. Then we can pass it down to this func. This local context can then be created like so:
```golang
ctx, cancel := context.WithCancel(parentCtx)
```

This has the added benefit of not having to deal with timeout calculation between all leaf `JitterUntil` calls. The local context is still needed to stop `JitterUntil` from spinning.

**Review by rosti — COMMENTED** [review]
Thanks @neolit123 !

File: cmd/kubeadm/app/discovery/token/token.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/discovery/token/token.go]: my initial impression is that this is not needed, but i will double check it.

Comment (neolit123) [reply cmd/kubeadm/app/discovery/token/token.go]: > This has the added benefit of not having to deal with timeout calculation between all leaf JitterUntil calls. The local context is still needed to stop JitterUntil from spinning.

unless i'm wrong this is problematic. we want a new context with timeout calculated each time getClusterInfo() is called.

https://golang.org/src/context/context.go?s=14448:14525#L453
```
func WithTimeout(parent Context, timeout time.Duration) (Context, CancelFunc) {
	return WithDeadline(parent, time.Now().Add(timeout))
```

i'd prefer we keep the code the way it is.

for the current use of JitterUntil we technically don't need the function to accept a channel, but a "duration" argument like Poll* would have been fine too.

Comment (rosti) [reply cmd/kubeadm/app/discovery/token/token.go]: Ok, let's keep things like this.
/lgtm

**Pull request merged**